### PR TITLE
Blocking Mode for pattern

### PR DESCRIPTION
**Pull request by InusualZ**
Urgent review!
* Variable Names
* Localization Names
* Method Description

### CODE REVIEW

File: src/main/java/com/raoulvdberge/refinedstorage/apiimpl/autocrafting/CraftingPattern.java [hunk 1]
```diff
@@ -32,6 +32,7 @@
     private List<ItemStack> outputs = new ArrayList<>();
     private List<ItemStack> byproducts = new ArrayList<>();
     private boolean mekanism = false; // Cause they are special in so many ways ¯\_(ツ)_/¯
+    private boolean blockingTask = false;
```
Comment (raoulvdberge) [inline src/main/java/com/raoulvdberge/refinedstorage/apiimpl/autocrafting/CraftingPattern.java]: what's the use of this variable?

File: src/main/resources/assets/refinedstorage/lang/en_US.lang [hunk 2]
```diff
@@ -101,6 +102,7 @@ misc.refinedstorage:set=Set
 misc.refinedstorage:cancel_all=Cancel All
 misc.refinedstorage:priority=Priority
 misc.refinedstorage:oredict=Oredict
+misc.refinedstorage:blocking_task=Blocking Mode
```
Comment (raoulvdberge) [inline src/main/resources/assets/refinedstorage/lang/en_US.lang]: we should just have one language key, that just says "Blocking"

**Comment by InusualZ** [comment]
Updated with a new approach. I'm have encountered a bug, but I don't know how to fix or what I have done wrong. When I restart the client the system doesn't recognize the pattern.

Picture of the result:
pattern-encoder
crafting-monitor
pattern-grid

PD: I don't think the pattern grid blocking feature is needed.

**Comment by raoulvdberge** [comment]
Yeah, I agree. Pattern grid doesn't need it I think. So, 2 questions:

- What do you mean with "When I restart the client, the system doesn't recognize the pattern."
- Does it work if a pattern that isn't blocked needs an item that IS blocked. Does that work correctly in the chain?

File: src/main/java/com/raoulvdberge/refinedstorage/api/autocrafting/task/ICraftingStep.java [hunk 3]
```diff
@@ -97,4 +97,9 @@
      * @return the written tag
      */
     NBTTagCompound writeToNBT(NBTTagCompound tag);
+
+    /**
+     * @return true weather if this step is blocked by another task
+     */
+    boolean isBlocked();
```
Comment (way2muchnoise) [inline src/main/java/com/raoulvdberge/refinedstorage/api/autocrafting/task/ICraftingStep.java]: Do we really need this?

File: src/main/java/com/raoulvdberge/refinedstorage/apiimpl/autocrafting/task/CraftingTask.java [hunk 4]
```diff
@@ -348,11 +354,17 @@ public boolean update(Map<ICraftingPatternContainer, Integer> usedContainers) {
...
 
             if (timesUsed++ <= container.getSpeedUpdateCount()) {
-                if (!step.hasStartedProcessing() && step.canStartProcessing(oreDictPrepped, networkFluids)) {
-                    step.setStartedProcessing();
-                    step.execute(toInsertItems, toInsertFluids);
-                    usedContainers.put(container, timesUsed);
-                    network.markCraftingMonitorForUpdate();
+                if ((container.isBlocked() && !step.getPattern().isProcessing()) || !container.isBlocked()) {
+                    if (!step.hasStartedProcessing() && step.canStartProcessing(oreDictPrepped, networkFluids)) {
+                        if (step.getPattern().isBlockingPattern()) {
```
Comment (way2muchnoise) [inline src/main/java/com/raoulvdberge/refinedstorage/apiimpl/autocrafting/task/CraftingTask.java]: Might want to move this to `step.setStartedProcessing()`

File: src/main/java/com/raoulvdberge/refinedstorage/apiimpl/autocrafting/task/CraftingTask.java [hunk 5]
```diff
@@ -535,6 +547,8 @@ public NBTTagCompound writeToNBT(NBTTagCompound tag) {
                             element = new CraftingMonitorElementError(element, "gui.refinedstorage:crafting_monitor.machine_none");
                         } else if (!step.hasStartedProcessing() && !step.canStartProcessing()) {
                             element = new CraftingMonitorElementError(element, "gui.refinedstorage:crafting_monitor.machine_in_use");
+                        } else if (!step.hasStartedProcessing() && step.isBlocked()) {
+                            element = new CraftingMonitorElementError(element, "gui.refinedstorage:crafting_monitor.blocked");
```
Comment (way2muchnoise) [inline src/main/java/com/raoulvdberge/refinedstorage/apiimpl/autocrafting/task/CraftingTask.java]: I can see why you would need `step.isBlocked()` but is `Machine in use` not kinda the same? imo not really needed. paging @raoulvdberge for this one

**Review by way2muchnoise — CHANGES_REQUESTED** [review]
Start looking pretty good :+1:

File: src/main/java/com/raoulvdberge/refinedstorage/apiimpl/autocrafting/task/CraftingTask.java (hunk 5 above)
Comment (raoulvdberge) [inline src/main/java/com/raoulvdberge/refinedstorage/apiimpl/autocrafting/task/CraftingTask.java]: nah this is fine. they're 2 different things

File: src/main/java/com/raoulvdberge/refinedstorage/apiimpl/network/node/NetworkNodeCrafter.java [hunk 6]
```diff
@@ -204,4 +212,14 @@ public IItemHandler getDrops() {
...
     }
+
+    @Override
+    public boolean isBlocked() {
+        return blocked;
+    }
+
+    @Override
+    public void setBlocked(boolean value) {
+        blocked = value;
```
Comment (raoulvdberge) [inline src/main/java/com/raoulvdberge/refinedstorage/apiimpl/network/node/NetworkNodeCrafter.java]: call markDirty here

File: src/main/java/com/raoulvdberge/refinedstorage/api/autocrafting/ICraftingPattern.java [hunk 7]
```diff
@@ -36,6 +36,11 @@
     boolean isOredict();
 
     /**
+     * @return true if the crafting pattern may block crafting step
+     */
+    boolean isBlockingPattern();
```
Comment (raoulvdberge) [inline src/main/java/com/raoulvdberge/refinedstorage/api/autocrafting/ICraftingPattern.java]: rename to isBlocking

File: src/main/java/com/raoulvdberge/refinedstorage/apiimpl/autocrafting/task/CraftingTask.java (hunk 5 above)
Comment (way2muchnoise) [inline src/main/java/com/raoulvdberge/refinedstorage/apiimpl/autocrafting/task/CraftingTask.java]: This should then be `step.getPattern().getContainer().isBlocked()` which eliminates the use of `ICraftingStep#isBlocked()`

File: src/main/java/com/raoulvdberge/refinedstorage/apiimpl/autocrafting/task/CraftingTask.java [hunk 8]
```diff
@@ -348,11 +354,13 @@ public boolean update(Map<ICraftingPatternContainer, Integer> usedContainers) {
             }
 
             if (timesUsed++ <= container.getSpeedUpdateCount()) {
-                if (!step.hasStartedProcessing() && step.canStartProcessing(oreDictPrepped, networkFluids)) {
-                    step.setStartedProcessing();
-                    step.execute(toInsertItems, toInsertFluids);
-                    usedContainers.put(container, timesUsed);
-                    network.markCraftingMonitorForUpdate();
+                if ((container.isBlocked() && !step.getPattern().isProcessing()) || !container.isBlocked()) {
```
Comment (way2muchnoise) [inline src/main/java/com/raoulvdberge/refinedstorage/apiimpl/autocrafting/task/CraftingTask.java]: Should be `!step.getPattern().isProcessing() || !container.isBlocked()`

**Review by way2muchnoise — APPROVED** [review]

**Pull request merged**